File: jpa/spring-data-jpa/jpashop-java/src/main/java/jpastudy/jpashop/repository/MemberRepository.java
package jpastudy.jpashop.repository;

import jpastudy.jpashop.domain.Member;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.PersistenceContext;
import javax.persistence.PersistenceUnit;
import java.util.List;

@Repository
@RequiredArgsConstructor
public class MemberRepository {

//    // 해당 에노테이션이 있으면 엔티티매니저를 스프링이 생성한 em을 여기에 주입해준다.
//    @PersistenceContext
//    private EntityManager em;

    private final EntityManager em;

    public void save(Member member) {
        em.persist(member);
    }

    public Member findOne(Long id) {
        return em.find(Member.class, id);
    }

    // sql은 테이블을 대상으로 쿼리를 날리지만, jpql은 엔티티를 대상으로 쿼리를 날린다.
    public List<Member> findAll() {
        return em.createQuery("select m from Member m", Member.class)
                .getResultList();
    }

    public List<Member> findByName(String name) {
        return em.createQuery("select m from Member m where m.name = :name")
                .setParameter("name", name)
                .getResultList();
    }
}

File: spring-data-jpa/kotjparelation/src/main/kotlin/hello/study/kotjparelation/repository/MemberRepository.kt
package hello.study.kotjparelation.repository

import hello.study.kotjparelation.domain.Member
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManager

@Repository
@Transactional
class MemberRepository(val em: EntityManager) {

    fun save(member: Member): Long? {
        em.persist(member)
        return member.id
    }

    fun findByName(name: String): List<Member> {
        return em.createQuery("select m from Member m where m.name=:name",Member::class.java)
            .setParameter("name", name)
            .resultList
    }

    fun findOne(id: Long): Member {
        return em.find(Member::class.java, id)
    }
}
